platform/ui/src/contextProviders/getPresentationIds.ts

```diff
@@ -0,0 +1,114 @@
+// a single display set in two or more viewports, when the user has simply
+// dragged and dropped the view in twice.  For example, it allows displaying
+// bone, brain and soft tissue views of a single display set, and to still
+// remember the specific changes to each viewport.
+const addUniqueIndex = (arr, key, viewports) => {
+  arr.push(0);
+  for (let displayInstance = 0; displayInstance < 128; displayInstance++) {
+    arr[arr.length - 1] = displayInstance;
+    const testId = arr.join(JOIN_STR);
+    if (
+      !viewports.find(
+        it => it.viewportOptions?.presentationIds?.[key] === testId
+      )
+    ) {
+      break;
+    }
+  }
+};
+
+const getLutId = (ds): string => {
+  if (!ds || !ds.options) return 'default';
+  if (ds.options.id) return ds.options.id;
+  const arr = Object.entries(ds.options).map(([key, val]) => `${key}=${value}`);
+  if (!arr.length) return 'default';
+  return arr.join('&');
+};
+
+export type PresentationIds = {
+  positionPresentationId?: string;
+  lutPresentationId?: string;
+};
+
+/**
+ * Gets a set of presentation IDs for a viewport.  The presentation IDs are
+ * used to remember the presentation state of the viewport when it is navigated
+ * to different layouts.
+ *
+ * The design of this is setup to allow preserving the view information in the
+ * following cases:
+ *
+ *
+ *      * If a set of display sets was previously displayed in the same initial
+ *        position as it is currently being asked to be displayed,
+ *        then remember the camera position as previously displayed
+ *
+ *      * If a set of display sets was previously displayed with the same initial
+ *        LUT conditions, then remember the last LUT displayed for that display set
+ *        and re-apply it.
+ *
+ *      * Otherwise, apply the initial hanging protocol specified LUT and camera
+ *        position to new display sets.
+ *
+ * This means generating two presentationId keys:
+ *
+ *  `positionPresentationId`
+ *
+ *  Used for getting the camera/initial position state sync values.
+ *  This is a combination of:
+ *      * `viewportOptions.id`
+ *      * `viewportOptions.orientation`
+ *      * display set UID's - as displayed for this viewport, excluding seg
+ *      * a unique index number if the previous key is already displayed
+ *
+ * `lutPresentationId`
+ *
+ * Used for getting the voi LUT information.  Generated from:
+ *
+ *       * `displaySetOption[0].options` - including the id if present
+ *       * displaySetUID's
+ *       * a unique index number if the previously generated key is already
+ *         displayed.
+ *
+ * @param viewport requiring a presentation Id
+ * @param viewports is the list of viewports being shown.  Any presentation ID's
+ *         among them must not be re-used in order to have each viewport have it's own presentation ID.
+ * @returns PresentationIds
+ */
+const getPresentationId = (viewport, viewports): PresentationIds => {
```

Comment:
can we type the parameters please?

Reply:
No, they don't have existing types aside from the ones in Cornerstone extension.  I agree it would be nice to have, but without support from the ViewportService, it isn't easy to add here.